bool isStart
bool isEnd
int32 idx